# Mini-Project - Text Learning

##### Student Tags

Author: Anderson Hitoshi Uyekita    
Mini-Project: Text Learning  
Course: Data Science - Foundations II  
COD: ND111  
Date: 21/01/2019    

***

## Table of Contents
- [Introduction](#intro)
- [Given code 1](#code1)
- [Exercise 1](#part_i_1)
- [Exercise 2](#part_i_2)
- [Exercise 3](#part_i_3)
- [Exercise 4](#part_i_4)
- [Exercise 5](#part_i_5)
- [Exercise 6](#part_i_6)

***

## General Information

This Jupyter Notebook (in Python 2) aims to create a reproducible archive.

## Introduction <a id='intro'></a>

In the beginning of this class, you identified emails by their authors using a number of supervised classification algorithms. In those projects, we handled the preprocessing for you, transforming the input emails into a TfIdf so they could be fed into the algorithms. Now you will construct your own version of that preprocessing step, so that you are going directly from raw data to processed features.

You will be given two text files: one contains the locations of all the emails from Sara, the other has emails from Chris. You will also have access to the parseOutText() function, which accepts an opened email as an argument and returns a string containing all the (stemmed) words in the email.

### Pre - Mini Project

In [1]:
# Importing nltk
#import nltk

# Downloading the corpus.
#nltk.download() # Choose identifier: "all-corpora"

In [2]:
# Importing the package stopwords.
from nltk.corpus import stopwords

# Creating a stopword.
sw = stopwords.words("english")

In [3]:
sw[0]

u'i'

In [4]:
sw[1]

u'me'

In [5]:
sw[10]

u"you've"

In [6]:
print "Number of words in stopwords:", len(sw)

Number of words in stopwords: 179


## Mini Project

## Exercise 1 - Warming Up with parseOutText() <a id='part_i_1'></a>

You’ll start with a warmup exercise to get acquainted with parseOutText(). Go to the tools directory and run parse_out_email_text.py, which contains parseOutText() and a test email to run this function over.

parseOutText() takes the opened email and returns only the text part, stripping away any metadata that may occur at the beginning of the email, so what's left is the text of the message. We currently have this script set up so that it will print the text of the email to the screen.

In [7]:
#!/usr/bin/python

from nltk.stem.snowball import SnowballStemmer
import string

def parseOutText(f):
    """ given an opened email file f, parse out all text below the
        metadata block at the top
        (in Part 2, you will also add stemming capabilities)
        and return a string that contains all the words
        in the email (space-separated) 
        
        example use case:
        f = open("email_file_name.txt", "r")
        text = parseOutText(f)
        
        """

    f.seek(0)  ### go back to beginning of file (annoying)
    all_text = f.read()

    ### split off metadata
    content = all_text.split("X-FileName:")
    words = ""
    if len(content) > 1:
        ### remove punctuation
        text_string = content[1].translate(string.maketrans("", ""), string.punctuation)

        ### project part 2: comment out the line below
        words = text_string

        ### split the text string into individual words, stem each word,
        ### and append the stemmed word to words (make sure there's a single
        ### space between each stemmed word)
        
    return words

def main():
    ff = open("../text_learning/test_email.txt", "r")
    text = parseOutText(ff)
    print text

if __name__ == '__main__':
    main()



Hi Everyone  If you can read this message youre properly using parseOutText  Please proceed to the next part of the project



>**What is the text that you get when you run parseOutText()?**

Hi Everyone  If you can read this message youre properly using parseOutText  Please proceed to the next part of the project

## Exercise 2 - Deploying Stemming <a id='part_i_2'></a>

In parseOutText(), comment out the following line: 

`words = text_string`

Augment parseOutText() so that the string it returns has all the words stemmed using a SnowballStemmer (use the nltk package, some examples that I found helpful can be found here: http://www.nltk.org/howto/stem.html ). Rerun parse_out_email_text.py, which will use your updated parseOutText() function--what’s your output now?

Hint: you'll need to break the string down into individual words, stem each word, then recombine all the words into one string.

In [8]:
#!/usr/bin/python

from nltk.stem.snowball import SnowballStemmer
import string

def parseOutText_v2(f):
    """ given an opened email file f, parse out all text below the
        metadata block at the top
        (in Part 2, you will also add stemming capabilities)
        and return a string that contains all the words
        in the email (space-separated) 
        
        example use case:
        f = open("email_file_name.txt", "r")
        text = parseOutText(f)
        
    """

    f.seek(0)  ### go back to beginning of file (annoying)
    all_text = f.read()

    ### split off metadata
    content = all_text.split("X-FileName:")
    words = ""
    if len(content) > 1:
        ### remove punctuation
        text_string = content[1].translate(string.maketrans("", ""), string.punctuation)

        ### project part 2: comment out the line below
        words = text_string

        ### split the text string into individual words, stem each word,
        ### and append the stemmed word to words (make sure there's a single
        ### space between each stemmed word)
        
        # Converting \n in spaces
        words = words.replace('\n',' ')

        # Removing punctuation
        # https://stackoverflow.com/questions/265960/best-way-to-strip-punctuation-from-a-string-in-python
        words = words.translate(string.maketrans("",""), string.punctuation)
        
        # Importing re package
        import re
        
        # Removing any instance of double or more spaces.
        words = re.sub(' +', ' ',words)
        
        # Removing spaces in the begining and ending of the email.
        words = words.lstrip().rstrip()
        
        # Splitting by space. Creating a vector of words.
        words = words.split()
        
        # Creating the Classifier using english as language.
        stemmer = SnowballStemmer("english")
        
        # Stemming each word of the vector.
        words = map(lambda x : stemmer.stem(x), words);
        
        # Binding all words together to became a string again.
        words = ' '.join(words)
            
    # Converting to UTF-8
    return words.encode("utf-8")

def main():
    ff = open("../text_learning/test_email.txt", "r")
    text = parseOutText_v2(ff)
    print text

if __name__ == '__main__':
    main()

hi everyon if you can read this messag your proper use parseouttext pleas proceed to the next part of the project


>**Now that something is deployed in `parseOutText()`, what is the output from parse_out_email_text.py?**

hi everyon if you can read this messag your proper use parseouttext pleas proceed to the next part of the project

## Exercise 3 - Clean Away "Signature Words" <a id='part_i_3'></a>

In vectorize_text.py, you will iterate through all the emails from Chris and from Sara. For each email, feed the opened email to parseOutText() and return the stemmed text string. Then do two things:

remove signature words (“sara”, “shackleton”, “chris”, “germani”--bonus points if you can figure out why it's "germani" and not "germany")
append the updated text string to word_data -- if the email is from Sara, append 0 (zero) to from_data, or append a 1 if Chris wrote the email.
Once this step is complete, you should have two lists: one contains the stemmed text of each email, and the second should contain the labels that encode (via a 0 or 1) who the author of that email is.

Running over all the emails can take a little while (5 minutes or more), so we've added a temp_counter to cut things off after the first 200 emails. Of course, once everything is working, you'd want to run over the full dataset.

In the box below, put the string that you get for word_data[152].

#### Given Code 1 <a id='code1'></a>

In [9]:
#!/usr/bin/python

import os
import pickle
import re
import sys
import time


"""
    Starter code to process the emails from Sara and Chris to extract
    the features and get the documents ready for classification.

    The list of all the emails from Sara are in the from_sara list
    likewise for emails from Chris (from_chris)

    The actual documents are in the Enron email dataset, which
    you downloaded/unpacked in Part 0 of the first mini-project. If you have
    not obtained the Enron email corpus, run startup.py in the tools folder.

    The data is stored in lists and packed away in pickle files at the end.
"""

from_sara  = open("from_sara.txt", "r")
from_chris = open("from_chris.txt", "r")

from_data = []
word_data = []

### temp_counter is a way to speed up the development--there are
### thousands of emails from Sara and Chris, so running over all of them
### can take a long time
### temp_counter helps you only look at the first 200 emails in the list so you
### can iterate your modifications quicker
temp_counter = 0

# Start
start = time.time()

for name, from_person in [("sara", from_sara), ("chris", from_chris)]:
    for path in from_person:
        ### only look at first 200 emails when developing
        ### once everything is working, remove this line to run over full dataset
        temp_counter += 1
        if temp_counter < 200:
            path = os.path.join('..', path[:-1])
            #print path
            email = open(path, "r")

            ### use parseOutText to extract the text from the opened email
            parsed_email = parseOutText_v2(email)
            
            # Converting spaces in spaces
            parsed_email = parsed_email.replace('\n',' ')
            
            # Importing re
            import re
        
            # Removing any instance of double or more spaces.
            parsed_email = re.sub(' +', ' ',parsed_email)
                
            ### use str.replace() to remove any instances of the words
            ### ["sara", "shackleton", "chris", "germani"]
            parsed_email_clean = parsed_email.replace('sara','').replace('chris','').replace('shackleton','').replace('germani','')
     
            ### append the text to word_data
            word_data.append(parsed_email_clean)
            
            ### append a 0 to from_data if email is from Sara, and 1 if email is from Chris
            if name == 'sara':
                from_data.append(0)
            elif name == 'chris':
                from_data.append(1)
            else:
                print "ERROR!!"
            
            # Converting to UTF-8
            word_data = map(lambda x : x.encode("utf-8"), word_data)
            
            email.close()

print "emails processed"
from_sara.close()
from_chris.close()

pickle.dump( word_data, open("your_word_data.pkl", "w") )
pickle.dump( from_data, open("your_email_authors.pkl", "w") )

# Stop
stop = time.time()

print "Processing time: ", round((stop - start)/60,2), "minutes"

emails processed
Processing time:  0.01 minutes


In [10]:
word_data[152]

'tjonesnsf stephani and sam need nymex calendar'

## Exercise 4 - TfIdf It" <a id='part_i_4'></a>

Transform the word_data into a tf-idf matrix using the sklearn TfIdf transformation. Remove english stopwords.

You can access the mapping between words and feature numbers using get_feature_names(), which returns a list of all the words in the vocabulary.

In [11]:
#!/usr/bin/python

import os
import pickle
import re
import sys
import time


"""
    Starter code to process the emails from Sara and Chris to extract
    the features and get the documents ready for classification.

    The list of all the emails from Sara are in the from_sara list
    likewise for emails from Chris (from_chris)

    The actual documents are in the Enron email dataset, which
    you downloaded/unpacked in Part 0 of the first mini-project. If you have
    not obtained the Enron email corpus, run startup.py in the tools folder.

    The data is stored in lists and packed away in pickle files at the end.
"""

from_sara  = open("from_sara.txt", "r")
from_chris = open("from_chris.txt", "r")

from_data = []
word_data = []

### temp_counter is a way to speed up the development--there are
### thousands of emails from Sara and Chris, so running over all of them
### can take a long time
### temp_counter helps you only look at the first 200 emails in the list so you
### can iterate your modifications quicker
temp_counter = 0

# Start
start = time.time()

for name, from_person in [("sara", from_sara), ("chris", from_chris)]:
    for path in from_person:
        ### only look at first 200 emails when developing
        ### once everything is working, remove this line to run over full dataset
        temp_counter += 1
        if temp_counter > -1:
            path = os.path.join('..', path[:-1])
            #print path
            email = open(path, "r")

            ### use parseOutText to extract the text from the opened email
            parsed_email = parseOutText_v2(email)
                
            ### use str.replace() to remove any instances of the words
            ### ["sara", "shackleton", "chris", "germani"]
            parsed_email_clean = parsed_email.replace('sara','').replace('chris','').replace('shackleton','').replace('germani','')
                      
            ### append the text to word_data
            word_data.append(parsed_email_clean)
            
            ### append a 0 to from_data if email is from Sara, and 1 if email is from Chris
            if name == 'sara':
                from_data.append(0)
            elif name == 'chris':
                from_data.append(1)
            else:
                print "ERROR!!"
            
            email.close()

print "emails processed"
from_sara.close()
from_chris.close()

pickle.dump( word_data, open("your_word_data.pkl", "w") )
pickle.dump( from_data, open("your_email_authors.pkl", "w") )

# Stop
stop = time.time()

print "Processing time: ", round((stop - start)/60,2), "minutes"

emails processed
Processing time:  0.99 minutes


In [12]:
### in Part 4, do TfIdf vectorization here

# Importing the TfIdf package.
from sklearn.feature_extraction.text import TfidfVectorizer

# Creating the vectorizer using english stopwords.
vectorizer = TfidfVectorizer(stop_words = 'english', lowercase=True)

# Fitting and Transforming.
vectorizer.fit_transform(word_data)

# Creating the Vocabulary List.
vocab_list = vectorizer.get_feature_names()

# Length of the Vocabulary.
len(vectorizer.get_feature_names())

38757

>** How many different words are there?**

38757

## Exercise 5 - TfIdf It" <a id='part_i_5'></a>

>**What is word number 34597 in your TfIdf?**

(Just to be clear--if the question were "what is word number 100," we would be looking for the word corresponding to vocab_list[100]. Zero-indexed arrays are so confusing to talk about sometimes.)

In [13]:
vocab_list[34597].encode('utf-8')

'stephaniethank'

#### Copying file

In [14]:
# Importing shutil to deal with copy
from shutil import copyfile

# File name
filename = 'vectorize_text.ipynb'

# Lesson
lesson = '11-Lesson_11'

# Directory to make a copy
dir_copy = '../../' + lesson + '/00-Mini Project/' + filename

# Copying file.
copyfile(filename, dir_copy)